packages/@aws-cdk/aws-synthetics/lib/code.ts

```diff
@@ -127,18 +128,25 @@ export class AssetCode extends Code {
    *
    * @param handler the canary handler
    */
-  private validateCanaryAsset(handler: string, family: RuntimeFamily) {
+  private validateCanaryAsset(scope: Construct, handler: string, family: RuntimeFamily) {
+    // Get the staged (or copied) asset path.
+    const assetOutdir = cdk.Stage.of(scope)?.assetOutdir;
+    let assetPath = this.assetPath;
+    if (assetOutdir) {
+      assetPath = path.join(assetOutdir, this.asset!.assetPath);
+    }
+
     if (path.extname(this.assetPath) !== '.zip') {
-      if (!fs.lstatSync(this.assetPath).isDirectory()) {
+      if (this.asset!.isFile) {
```

Comment:
I understand that with the current ordering of things, `this.asset` exists, but this is a detail that we should abstract out of the function. I prefer if we never use `!` and instead just make sure that `this.asset` exists

```ts
if (this.asset && this.asset.isFile) {}
```